src/coreclr/src/vm/genericdict.cpp

```diff
@@ -640,6 +768,105 @@ Dictionary::Restore(
+    if (pIMD->GetDictionarySlotsSize() != sizeFromDictLayout)
+    {
+        _ASSERT(pIMD->GetDictionarySlotsSize() < sizeFromDictLayout);
+
+        pDictionary = (Dictionary*)(void*)pIMD->GetLoaderAllocator()->GetHighFrequencyHeap()->AllocMem(S_SIZE_T(sizeFromDictLayout + sizeof(void*)));
+        ZeroMemory(pDictionary, sizeFromDictLayout + sizeof(void*));
+
+        // Slot[-1] points at previous dictionary to help with diagnostics when investigating crashes
+        *(byte**)pDictionary = (byte*)pIMD->m_pPerInstInfo.GetValue() + 1;
+        pDictionary++;
+
+        // Copy old dictionary entry contents
+        memcpy(pDictionary, (const void*)pIMD->m_pPerInstInfo.GetValue(), pIMD->GetDictionarySlotsSize());
+
+        ULONG_PTR* pSizeSlot = ((ULONG_PTR*)pDictionary) + pIMD->GetNumGenericMethodArgs();
+        *pSizeSlot = sizeFromDictLayout;
+
+        // Flush any write buffers before publishing the new dictionary contents
+        FlushProcessWriteBuffers();
+
+        // Publish the new dictionary slots to the type.
+        FastInterlockExchangePointer((TypeHandle**)pIMD->m_pPerInstInfo.GetValuePtr(), (TypeHandle*)pDictionary);
+    }
+#endif
+
+    RETURN pDictionary;
+}
+
+Dictionary* Dictionary::GetTypeDictionaryWithSizeCheck(MethodTable* pMT)
+{
+    CONTRACT(Dictionary*)
+    {
+       THROWS;
+       GC_TRIGGERS;
+       PRECONDITION(SystemDomain::SystemModule()->m_DictionaryCrst.OwnedByCurrentThread());
+       POSTCONDITION(CheckPointer(RETVAL));
+    }
+    CONTRACT_END;
+
+    Dictionary* pDictionary = pMT->GetDictionary();
+
+#if !defined(CROSSGEN_COMPILE)
+    DictionaryLayout* pDictLayout = pMT->GetClass()->GetDictionaryLayout();
+    _ASSERTE(pDictLayout != NULL && pDictLayout->GetMaxSlots() > 0);
+
+    DWORD sizeFromDictLayout = DictionaryLayout::GetDictionarySizeFromLayout(pMT->GetNumGenericArgs(), pDictLayout);
+    if (pMT->GetDictionarySlotsSize() != sizeFromDictLayout)
+    {
+        _ASSERTE(pMT->GetDictionarySlotsSize() < sizeFromDictLayout);
+
+        // Expand type dictionary
+        pDictionary = (Dictionary*)(void*)pMT->GetLoaderAllocator()->GetHighFrequencyHeap()->AllocMem(S_SIZE_T(sizeFromDictLayout + sizeof(void*)));
+        ZeroMemory(pDictionary, sizeFromDictLayout + sizeof(void*));
+
+        // Slot[-1] points at previous dictionary to help with diagnostics when investigating crashes
+        *(byte**)pDictionary = (byte*)pMT->GetPerInstInfo()[pMT->GetNumDicts() - 1].GetValue() + 1;
+        pDictionary++;
+
+        // Copy old dictionary entry contents
+        memcpy(pDictionary, (const void*)pMT->GetPerInstInfo()[pMT->GetNumDicts() - 1].GetValue(), pMT->GetDictionarySlotsSize());
+
+        ULONG_PTR* pSizeSlot = ((ULONG_PTR*)pDictionary) + pMT->GetNumGenericArgs();
+        *pSizeSlot = sizeFromDictLayout;
+
+        // Flush any write buffers before publishing the new dictionary contents
+        FlushProcessWriteBuffers();
```

Comment:
This should not be needed with the simpler scheme.

Reply:
It might be possible that thread A allocates a new dictionary, copies the entries from the old dictionary, and update the m_pPerInstInfo pointer to that new dictionary, and thread B sees the new pointer, and attempts to read slots from that new dictionary, but gets NULL/invalid values from the cpu pipeline because the memcpy hasn't fully flushed the changes. Or am I misunderstanding something?

Reply:
> gets NULL/invalid values from the cpu pipeline 

That cannot happen because of [data dependency ordering](https://www.bing.com/search?q=Data+Dependency+Ordering). The read of the dictionary slot depends on the read of the pointer to the dictionary. The data dependency ordering is enforced all hardware that we care about. We depend on it in many places.

Reply:
Note that `FlushProcessWriteBuffers` is pretty expensive operation. It sends interrupt to all cores. You want to use it only when you really have to.

Reply:
Ok thanks for the clarification Jan and confirming that the invalid state can no longer happen. I will remove this flushing operation.